ecosystem/indexer/src/indexer/tailer.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    database::PgDbPool,
+    indexer::{
+        errors::TransactionProcessingError, fetcher::TransactionFetcher,
+        processing_result::ProcessingResult, transaction_processor::TransactionProcessor,
+    },
+};
+use aptos_logger::info;
+use aptos_rest_client::Transaction;
+use std::{fmt::Debug, sync::Arc};
+use tokio::{sync::Mutex, task::JoinHandle};
+use url::{ParseError, Url};
+
+diesel_migrations::embed_migrations!();
+
+#[derive(Clone)]
+pub struct Tailer {
+    transaction_fetcher: Arc<Mutex<TransactionFetcher>>,
+    processors: Vec<Arc<dyn TransactionProcessor>>,
+    connection_pool: PgDbPool,
+}
+
+impl Tailer {
+    pub fn new(node_url: &str, connection_pool: PgDbPool) -> Result<Tailer, ParseError> {
+        let url = Url::parse(node_url)?;
+        let transaction_fetcher = TransactionFetcher::new(url, None);
+        Ok(Self {
+            transaction_fetcher: Arc::new(Mutex::new(transaction_fetcher)),
+            processors: vec![],
+            connection_pool,
+        })
+    }
+
+    pub fn run_migrations(&self) {
+        info!("Running migrations...");
+        embedded_migrations::run_with_output(
+            &self
+                .connection_pool
+                .get()
+                .expect("Could not get connection for migrations"),
+            &mut std::io::stdout(),
+        )
+        .expect("migrations failed!");
+        info!("Migrations complete!");
+    }
+
+    pub fn add_processor(&mut self, processor: Arc<dyn TransactionProcessor>) {
+        info!("Adding processor to indexer: {}", processor.name());
+        self.processors.push(processor);
+    }
+
+    /// For all versions which have an `ok=false` in the `processor_status` table, re-run them
+    pub async fn handle_previous_errors(&self) {
```

Comment:
when would we get here with the expect statements?

Reply:
this is only run on startup